File: tools/kubexpose.rb
#! /usr/bin/ruby

module KubeExpose
    def KubeExpose.newport()
        return 31000 + Random.rand(1000)
    end

    def KubeExpose.expose(servicename, port=newport())
        `kubectl patch service #{servicename} --type='json' \
        -p='[{"op": "replace", "path": "/spec/type", "value": "NodePort"},
            {"op": "replace", "path": "/spec/ports/0/nodePort", "value": "#{port}"}]'`
        return (if $?.success? then port else 0 end)
    end

    def KubeExpose.hide(servicename)
        `kubectl patch service #{servicename} --type='json' \
        -p='[{"op": "replace", "path": "/spec/type", "value": "ClusterIP"},
            {"op": "remove", "path": "/spec/ports/0/nodePort"}]'`
        return $?.success?
    end
end

if __FILE__ == $0
    if ARGV.length != 2 || (ARGV[1] != 'expose' && ARGV[1] != 'hide')
        puts "Usage: kubexpose.rb <servicename> <expose/hide>"
        exit 1
    end
    servicename = ARGV[0]
    action = ARGV[1]

    if action == "expose"
        res = KubeExpose.expose(servicename)
        if res != 0 then puts res end
    else 
        res = KubeExpose.hide(servicename)
    end
end

File: tools/kubexpose/kubexpose.py
#! /usr/bin/python

from random import randint
import sys
from subprocess import Popen, PIPE

def newport():
    return randint(31000, 32000)

def expose(servicename, port=newport()):
    cmd = 'kubectl patch service %s --type=\'json\' ' % servicename
    cmd += '-p=\'[{"op": "replace", "path": "/spec/type", "value": "NodePort"},'
    cmd += '{"op": "replace", "path": "/spec/ports/0/nodePort", "value": "%d"}]\'' % port
    output, err = Popen(cmd, shell=True, executable='/bin/bash', stdout=PIPE, stderr=PIPE).communicate()
    if err != None and err != "":
        return None
    return port

def hide(servicename):
    cmd = 'kubectl patch service %s --type=\'json\' ' % servicename
    cmd += '-p=\'[{"op": "replace", "path": "/spec/type", "value": "ClusterIP"},'
    cmd += '{"op": "remove", "path": "/spec/ports/0/nodePort"}]\''
    return Popen(cmd, shell=True, executable='/bin/bash', stdout=PIPE, stderr=PIPE).communicate()

if __name__ == "__main__":
    if len(sys.argv) != 3 or (sys.argv[2] != 'expose' and sys.argv[2] != 'hide'):
        print "Usage: kubexpose.rb <servicename> <expose/hide>"
        exit(0)

    servicename = sys.argv[1]
    action = sys.argv[2]

    if action == 'expose':
        port = expose(servicename)
        if port != None:
            print port
        else:
            print "Error exposing"
            exit(1)
    else:
        hide(servicename)
